tfx/orchestration/kubeflow/kubeflow_dag_runner.py

```diff
@@ -347,6 +350,13 @@ def _construct_pipeline_graph(self, pipeline: tfx_pipeline.Pipeline,
 
       component_to_kfp_op[component] = kfp_component.container_op
 
+    # If exit handler defined create an exit handler and add all ops to it.
+    if self._exit_handler:
+      exit_op = component_to_kfp_op[self._exit_handler]
```

Comment:
I might miss it, where does the exit_handler being converted to kfp ops?

How does FinalStatusStr get passed in to exit handler?

Reply:
It's being transformed as part of the main for loop that creates BaseComponent (since I added the exit_handler in the pipeline).

FinalStatusStr is what I need to figure out still. I think I can make it work but I need to modify a exec property value on the fly.

Reply:
Found how to replace FinalStatusStr. I used tfx/orchestration/kubeflow/utils.py to fix this. I basically just replace the class for a string that Argo will replace w status: `'{{workflow.status}}'`

See: https://github.com/argoproj/argo-workflows/blob/master/docs/variables.md#exit-handler for argo. 
Tekton has a different one: https://github.com/tektoncd/pipeline/blob/main/docs/variables.md#variables-available-in-a-pipeline

Reply:
Thanks for the pull.
The container_op converting seems right to me.

Has the logic about FinalStatusStr reflected in the pull yet? I didn't spot that.

It would be helpful to add the corresponding test with exit handler in /tfx/orchestration/kubeflow/kubeflow_dag_runner_test.py to validate the correctness.

Reply:
Thanks, sorry didn't notice the component is added to pipeline at line 413

Do you have a pic of the execution graph of an e2e pipeline with exit handler in kfp? we want to see the relation between node and see if it align with Vertex.

Reply:
@polydier1 yes, it has been updated. See utils.py to see where the logic for FinalStatusStr. Also added a test here (which passes).

I have a pic of a sample pipeline I was using w similar code internally (It may have different names since its not an exact copy from here, and its failed due to me messing up the docker image).

<img width="942" alt="Screen Shot 2021-12-16 at 8 20 41 PM" src="https://user-images.githubusercontent.com/6912589/146435201-f0447a6a-fbdb-4307-aa45-1bd50e45ba05.png">


<img width="677" alt="Screen Shot 2021-12-16 at 8 20 30 PM" src="https://user-images.githubusercontent.com/6912589/146435190-af9f7f08-470d-4ee2-b7ef-beb0f9838220.png">

Reply:
Interesting, in Vertex, the pipeline itself is treated as a whole, and exit handler will be executed even if pipeline failed to execute to the end.

Seems in this graph, exit_handler is treated as the downstream of filepusher? what if trainer throws an error and pusher doesn't get executed?

Reply:
it does get executed still. For some reason the UI shows it like this (see the other graph)

Reply:
Thank!!! I think this looks good in general!

(Will let ExitHandler owner take a look when she's back from vacation)

Reply:
sounds good! happy holidays y'all!

Reply:
let's double check the graph is correct before merging it. I have forwarded the picture to KFP team to verify.